src/main/kotlin/org/rust/cargo/project/settings/ui/RustProjectSettingsPanel.kt

```diff
@@ -102,43 +101,33 @@ class RustProjectSettingsPanel : JPanel() {
     private fun scheduleVersionUpdate(toolchainLocation: String) {
         versionUpdateAlarm.cancelAllRequests()
         versionUpdateAlarm.addRequest({
-            val toolchain = RustToolchain(toolchainLocation)
-
-            val rustc = toolchain.queryRustcVersion()
-            val cargo = toolchain.queryCargoVersion()
-            val rustup = toolchain.queryRustupVersion()
-
-            updateVersion(rustc, cargo, rustup)
-
+            val versionInfo = RustToolchain(toolchainLocation).queryVersions()
+            updateVersion(versionInfo)
         }, versionUpdateDelayMillis)
     }
 
-    private fun updateVersion(rustc: Version?, cargo: Version?, rustup: Version?) {
+    private fun updateVersion(info: RustToolchain.VersionInfo) {
         ApplicationManager.getApplication().invokeLater({
-            if (!Disposer.isDisposed(disposable)) {
-                if (rustc != null) {
-                    rustVersion.text        = rustc.release
-                    rustVersion.foreground  = JBColor.foreground()
+            if (Disposer.isDisposed(disposable)) return@invokeLater
+
+            val labelToVersion = listOf(
+                rustVersion to info.rustc,
+                cargoVersion to info.cargo,
+                rustupVersion to info.rustup
+            )
+
+            for ((label, version) in labelToVersion) {
+                if (version == null) {
+                    label.text = "N/A"
+                    label.foreground = JBColor.RED
                 } else {
-                    rustVersion.text        = "N/A"
-                    rustVersion.foreground  = JBColor.RED
-                }
-
-                if (cargo != null) {
-                    cargoVersion.text       = cargo.release
-                    cargoVersion.foreground = if (cargo >= RustToolchain.CARGO_LEAST_COMPATIBLE_VERSION) JBColor.foreground() else JBColor.RED
-                } else {
-                    cargoVersion.text       = "N/A"
-                    cargoVersion.foreground = JBColor.RED
+                    label.text = version.toString()
+                    label.foreground = JBColor.foreground()
                 }
+            }
 
-                if (rustup != null) {
-                    rustupVersion.text       = rustup.release
-                    rustupVersion.foreground = JBColor.foreground()
-                } else {
-                    rustupVersion.text       = "N/A"
-                    rustupVersion.foreground = JBColor.RED
-                }
+            if (!info.cargoHasMetadataCommand) {
+                cargoVersion.foreground = JBColor.RED
```

Comment:
I think it might be nice to provide user explicit reason why we don't like he's cargo.